source/host/KReplaySystem.cc

```diff
@@ -0,0 +1,209 @@
+
+    for (int i = 0; i < argc; ++i) {
+        std::optional<Host::EOption> flag = Host::Option::CheckFlag(argv[i]);
+        if (!flag || *flag == Host::EOption::Invalid) {
+            WARN("Expected a flag! Got: %s", argv[i]);
+            continue;
+        }
+
+        switch (*flag) {
+        case Host::EOption::Ghost: {
+            ASSERT(i + 1 < argc);
+
+            m_currentGhostFileName = argv[++i];
+            m_currentRawGhost = Abstract::File::Load(m_currentGhostFileName, m_currentRawGhostSize);
+
+            if (m_currentRawGhostSize < System::RKG_HEADER_SIZE ||
+                    m_currentRawGhostSize > System::RKG_UNCOMPRESSED_INPUT_DATA_SECTION_SIZE) {
+                PANIC("File cannot be a ghost! Check the file size.");
+            }
+
+            // Creating the raw ghost file validates it
+            System::RawGhostFile file = System::RawGhostFile(m_currentRawGhost);
+
+            m_currentGhost = new System::GhostFile(file);
+            ASSERT(m_currentGhost);
+        } break;
+        case Host::EOption::Invalid:
+        default:
+            PANIC("Invalid flag!");
+            break;
+        }
+    }
+}
+
+KReplaySystem *KReplaySystem::CreateInstance() {
+    ASSERT(!s_instance);
+    s_instance = new KReplaySystem;
+    return static_cast<KReplaySystem *>(s_instance);
+}
+
+void KReplaySystem::DestroyInstance() {
+    ASSERT(s_instance);
+    auto *instance = s_instance;
+    s_instance = nullptr;
+    delete instance;
+}
+
+KReplaySystem *KReplaySystem::Instance() {
+    return static_cast<KReplaySystem *>(s_instance);
+}
+
+KReplaySystem::KReplaySystem()
+    : m_currentGhostFileName(nullptr), m_currentGhost(nullptr), m_currentRawGhost(nullptr),
+      m_currentRawGhostSize(0) {}
+
+KReplaySystem::~KReplaySystem() {
+    if (s_instance) {
+        s_instance = nullptr;
+        WARN("KReplaySystem instance not explicitly handled!");
+    }
+
+    delete m_sceneMgr;
+    delete m_currentGhost;
+    delete m_currentRawGhost;
+}
+
+bool KReplaySystem::calcEnd() const {
+    constexpr u16 MAX_MINUTE_COUNT = 10;
+
+    const auto *raceManager = System::RaceManager::Instance();
+    if (raceManager->stage() == System::RaceManager::Stage::FinishGlobal) {
+        return true;
+    }
+
+    if (raceManager->timerManager().currentTimer().min >= MAX_MINUTE_COUNT) {
+        return true;
+    }
+
+    return false;
+}
+
+void KReplaySystem::reportFail(const char *msg) const {
+    std::string report(m_currentGhostFileName);
+    report += "\n" + std::string(msg);
+    Abstract::File::Append("results.txt", report.c_str(), report.size());
+}
+
+bool KReplaySystem::success() const {
+    const auto *raceManager = System::RaceManager::Instance();
+    if (raceManager->stage() != System::RaceManager::Stage::FinishGlobal) {
+        reportFail("Race didn't finish");
+        return false;
+    }
+
+    s32 desyncingTimerIdx = getDesyncingTimerIdx();
+    if (desyncingTimerIdx != -1) {
+        char msgBuffer[128];
```

Comment:
Any particular reason you are opting to use `snprintf` instead of a `std::string`?

Reply:
I do not like strings in C++, so I don't default to them. It can be done via `std::format`, but I didn't consider it at the time as it's a relatively modern feature.